flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp

```diff
@@ -155,28 +155,32 @@ mlir::LLVM::DITypeAttr DebugTypeGenerator::convertSequenceType(
     fir::SequenceType seqTy, mlir::LLVM::DIFileAttr fileAttr,
     mlir::LLVM::DIScopeAttr scope, mlir::Location loc) {
   mlir::MLIRContext *context = module.getContext();
-  // FIXME: Only fixed sizes arrays handled at the moment.
-  if (seqTy.hasDynamicExtents())
-    return genPlaceholderType(context);
 
   llvm::SmallVector<mlir::LLVM::DINodeAttr> elements;
   mlir::LLVM::DITypeAttr elemTy =
       convertType(seqTy.getEleTy(), fileAttr, scope, loc);
 
   for (fir::SequenceType::Extent dim : seqTy.getShape()) {
-    auto intTy = mlir::IntegerType::get(context, 64);
-    // FIXME: Only supporting lower bound of 1 at the moment. The
-    // 'SequenceType' has information about the shape but not the shift. In
-    // cases where the conversion originated during the processing of
-    // 'DeclareOp', it may be possible to pass on this information. But the
-    // type conversion should ideally be based on what information present in
-    // the type class so that it works from everywhere (e.g. when it is part
-    // of a module or a derived type.)
-    auto countAttr = mlir::IntegerAttr::get(intTy, llvm::APInt(64, dim));
-    auto lowerAttr = mlir::IntegerAttr::get(intTy, llvm::APInt(64, 1));
-    auto subrangeTy = mlir::LLVM::DISubrangeAttr::get(
-        context, countAttr, lowerAttr, nullptr, nullptr);
-    elements.push_back(subrangeTy);
+    if (dim == seqTy.getUnknownExtent()) {
+      auto subrangeTy = mlir::LLVM::DISubrangeAttr::get(
+          context, /*count=*/nullptr, /*lowerBound=*/nullptr,
+          /*upperBound*/ nullptr, /*stride*/ nullptr);
+      elements.push_back(subrangeTy);
```

Comment:
Will it be needed to make a difference between assumed-size and explicit shape with non constant extent here?

```
subroutine sub1(x, n)
 real :: x(n)
end subroutine
```

and:
```
subroutine sub1(x)
 real :: x(*)
end subroutine
```

Reply:
Thanks for catching this. We need to make a different between these 2 cases. For the 2nd case, we can generate a DISubrange pointing to a variable. But that will require DeclOp to be propagated to DebugTypeConvertor. This information is not available in Type only.

For this PR, I have added a FIXME and a testcase with variable extent arrays that is currently xfailed. I will also open an issue to keep track of this once this PR is merged.

Reply:
Won't reading from the descriptor at runtime work just as well in this case? Does the debug output look different for reading from the variable?

Reply:
As I see there is no descriptor in this case. The type of a variable I get is `!fir.array<?xf32>`

Reply:
Ok, we should probably add some attribute to tag assumed-size array xxx.declare (we can currently infer that looking at the last extent value that should be a -1 constant, but chain lookups are a bit weak IMHO, better to set an attribute.

> As I see there is no descriptor in this case.

To be more specific @tblah, there was one in HLFIR, but it is "unlink" when creating the fir.declare that has a single result. , and it may have been eliminated if unused in FIR.